### PR TITLE
feat: allow overriding SOA and NS DNS record TTL

### LINKED ISSUE #46: TTL is Terribly Low
## Describe the Feature

Increase the TTL on the Route53 resource records to something that will actually help DNS caches. 
 TTLs of 60 and 30 are way too low.  172800 tends to be more reasonable.

## Expected Behavior

TTL for DNS records that do not change very often should be measured in days or weeks.  SOA and NS records do not change after creation in Route53, so making DNS caches world-wide query for updated records after 30 seconds is a waste of bandwidth.

### PR DESCRIPTION
**Pull request by bartelemi**
## what
* Provide variables to configure TTL for SOA and NS records
* Original PR: Change default values of TTL for NS from 30 seconds to 2 days (172800 seconds)
* Original PR: Change default values of TTL for SOA from 60 seconds to 15 minutes (900 seconds)
* Preserve previously-hardcoded TTLs as defaults (@osterman @korenyoni) — see below
* Added the tenant context variable to the zone name template - maybe someone will need it :)

## why

Original PR: These values were hardcoded and with too small default values. DNS is supposed to be a cache, and having a low TTL of records like NS or SOA contradicts that idea.

The previously-hardcoded low TTLs are values preferred by Cloud Posse as low TTLs lead to better availability in the event of a negative DNS response, such that the negative DNS answer will not be cached for an extended period of time (see @osterman's comments in this thread).

AWS defaults for TTL:

NS  - 172800s
SOA - 900s

Clousposse's defaults:

NS  - 60s
SOA - 30s

## references
Closes #46

Previous PR has some weird permission issues which I couldn't solve, so I forked it again into a new repo and applied the same changes.

### CODE REVIEW

**Comment by bartelemi** [comment]
It seems that there are still some problems with cloudpossebot not being able to push the auto-format commit. Anyone can have a look what's causing it? I simply forked the repo and opened a PR - this process seems to work fine on your other github repositories...

**Comment by razorsedge** [comment]
Is there any reason you switched from 86400 to 900 for the SOA?

**Comment by bartelemi** [comment]
I had it at 86400 in the original PR, but I decided that it's better to stick with AWS default values:
https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/SOA-NSrecords.html

**Comment by Gowiem** [comment]
@bartelemi this looks like a good PR. Can you update to fix your conflicts and `tenant` to the new templating pattern?

Thanks!

**Comment by korenyoni** [comment]
Adding a `no-release` label because I want to release this alongside #52 and #51

**Review by korenyoni — COMMENTED** [review]
LGTM but please add tests.

If not I can take over this PR and add them myself.

**Comment by bartelemi** [comment]
@korenyoni I have extended the original _complete_ example with ttl overrides for both records. Hopefully that's enough, otherwise please let me know what else you're expecting.

File: variables.tf [hunk 1]
```diff
@@ -21,3 +21,15 @@ variable "parent_zone_record_enabled" {
   default     = true
   description = "Whether to create the NS record on the parent zone. Useful for creating a cluster zone across accounts. `var.parent_zone_name` required if set to false."
 }
+
+variable "ns_record_ttl" {
+  type        = number
+  default     = 172800
```
Comment (korenyoni) [inline variables.tf:27]:
```suggestion
  default     = 60
```

File: variables.tf [hunk 2]
```diff
@@ -21,3 +21,15 @@ variable "parent_zone_record_enabled" {
...
+  description = "The time to live (TTL) of the nameserver Route53 record, in seconds."
+}
+
+variable "soa_record_ttl" {
+  type        = number
+  default     = 900
```
Comment (korenyoni) [inline variables.tf:33]:
```suggestion
  default     = 30
```

File: variables.tf [hunk 3]
```diff
@@ -21,3 +21,15 @@ variable "parent_zone_record_enabled" {
   default     = true
   description = "Whether to create the NS record on the parent zone. Useful for creating a cluster zone across accounts. `var.parent_zone_name` required if set to false."
 }
+
+variable "ns_record_ttl" {
+  type        = number
+  default     = 172800
+  description = "The time to live (TTL) of the nameserver Route53 record, in seconds."
```
Comment (korenyoni) [inline variables.tf:28]:
```suggestion
  description = <<-EOT
  The time to live (TTL) of the nameserver Route53 record, in seconds.
  
  Note that the default value of `60` is very low, but is kept at that value to preserve backwards compatibility. 
  EOT
```

File: variables.tf [hunk 4]
```diff
@@ -21,3 +21,15 @@ variable "parent_zone_record_enabled" {
...
+}
+
+variable "soa_record_ttl" {
+  type        = number
+  default     = 900
+  description = "The time to live (TTL) of the start of authority Route53 record, in seconds."
```
Comment (korenyoni) [inline variables.tf:34]:
```suggestion
  description = <<-EOT
  The time to live (TTL) of the start of authority Route53 record, in seconds.
  
  Note that the default value of `30` is very low, but is kept at that value to preserve backwards compatibility. 
```

**Comment by bartelemi** [comment]
> I want to preserve the old TTL values as defaults for backwards compatibility.

As much as I understand the whole backwards compatibility reasoning, I don't think it's a major breaking change in this particular case. Original values were ridiculously low and it was based on nothing. The defaults I propose are based on official AWS documentation and I think aligning with that is much more important than "backwards compatibility" of some terraform module which doesn't even have a stable release yet.

Also, there is a reason why you include this comment everywhere in your modules:
```terraform
# Cloud Posse recommends pinning every module to a specific version
# version = "x.x.x"
```

That's just my 2 cents to the conversation, I don't want to stretch out this pull-request in time more than needed.

**Comment by korenyoni** [comment]
> > I want to preserve the old TTL values as defaults for backwards compatibility.
> 
> As much as I understand the whole backwards compatibility reasoning, I don't think it's a major breaking change in this particular case. Original values were ridiculously low and it was based on nothing. The defaults I propose are based on official AWS documentation and I think aligning with that is much more important than "backwards compatibility" of some terraform module which doesn't even have a stable release yet.
> 
> Also, there is a reason why you include this comment everywhere in your modules:
> 
> ```terraform
> # Cloud Posse recommends pinning every module to a specific version
> # version = "x.x.x"
> ```
> 
> That's just my 2 cents to the conversation, I don't want to stretch out this pull-request in time more than needed.

You're probably right — we are usually strict when it comes to backwards compatibility but this is a trivial case. I'll bring it up in an internal meeting in 20 minutes and get back to you.

File: variables.tf [hunk 5]
```diff
@@ -21,3 +21,25 @@ variable "parent_zone_record_enabled" {
...
+variable "ns_record_ttl" {
+  type        = number
+  default     = 60
+  description = <<-EOT
+  The time to live (TTL) of the nameserver Route53 record, in seconds.
+  
+  Note that the default value of `60` is very low, but is kept at that value to preserve backwards compatibility.
+  Higher values such as `172800` are suitable for nameserver records.
+  EOT
+}
```
Comment (korenyoni) [inline variables.tf:34]:
```suggestion
variable "ns_record_ttl" {
  type        = number
  default     = 172800
  description = <<-EOT
  The time to live (TTL) of the nameserver Route53 record, in seconds.
  
  The default value for AWS-managed NS records is `172800`.
  EOT
}
```

File: variables.tf [hunk 6]
```diff
@@ -21,3 +21,25 @@ variable "parent_zone_record_enabled" {
...
+variable "soa_record_ttl" {
+  type        = number
+  default     = 30
+  description = <<-EOT
+  The time to live (TTL) of the start of authority Route53 record, in seconds.
+  
+  Note that the default value of `30` is very low, but is kept at that value to preserve backwards compatibility.
+  Higher values such as `900` are suitable for SOA records. 
+  EOT
+}
```
Comment (korenyoni) [inline variables.tf:45]:
```suggestion
variable "soa_record_ttl" {
  type        = number
  default     = 900
  description = <<-EOT
  The time to live (TTL) of the start of authority Route53 record, in seconds.
  
  The default value for AWS-managed NS records is `900`.
  EOT
}
```

**Comment by osterman** [comment]
> Change default values of TTL for NS from 30 seconds to 2 days (172800 seconds)

I believe we can lengthen this, but high TTLs for DNS scare me. For context, having been at this for a long time, "It's always DNS" 

https://www.cscdbs.com/blog/why-dns-is-the-biggest-single-point-of-failure/
https://www.reddit.com/r/sysadmin/comments/puwt72/its_always_dns/
etc.

File: variables.tf [hunk 7]
```diff
@@ -21,3 +21,23 @@ variable "parent_zone_record_enabled" {
   default     = true
   description = "Whether to create the NS record on the parent zone. Useful for creating a cluster zone across accounts. `var.parent_zone_name` required if set to false."
 }
+
+variable "ns_record_ttl" {
+  type        = number
+  default     = 172800
```
Comment (korenyoni) [inline variables.tf:27]:
```suggestion
  default     = 60
```

**Review by korenyoni — APPROVED** [review]

**Comment by razorsedge** [comment]
@osterman Your comments on SOA are inaccurate.  SOA is meant to provide details for zone transfers between a domain's authoritative servers.  Yes, one of those values ([MINIMUM](https://datatracker.ietf.org/doc/html/rfc1035#section-3.3.13)) has been co-opted to now represent the TTL for NXDOMAIN, but we are not talking about that value here.  We are discussing the TTL of the SOA and NS records themselves.

High TTLs on SOA and NS have been the norm for DNS infrastructure for decades and I was rather taken aback that this module had set mine so low.

**Comment by Nuru** [comment]
> @osterman Your comments on SOA are inaccurate. SOA is meant to provide details for zone transfers between a domain's authoritative servers. Yes, one of those values ([MINIMUM](https://datatracker.ietf.org/doc/html/rfc1035#section-3.3.13)) has been co-opted to now represent the TTL for NXDOMAIN, but we are not talking about that value here. We are discussing the TTL of the SOA and NS records themselves.
> 
> High TTLs on SOA and NS have been the norm for DNS infrastructure for decades and I was rather taken aback that this module had set mine so low.

@razorsedge, as @osterman explained above, our previously hard coded settings, and now defaults, reflect our orientation towards working in DevOps and development environments. They are not tuned for production. Our typical workflow is to create resources using default parameters and then tune those parameters as the infrastructure stabilizes. Our defaults are not meant as, and should not be taken as, recommendations for production settings.

The point of the TTL of the SOA was never terribly well defined. Distribution of the SOA itself and the other records in a zone server is not controlled by the SOA TTL, but by the other fields of the SOA: `REFRESH`, `RETRY`, and `EXPIRE`. [RFC 1035](https://datatracker.ietf.org/doc/html/rfc1035#section-3.3.13) from 1987 says "SOA records are always distributed with a zero TTL to prohibit caching". In 1997, 10 years later, [RFC 2181](https://datatracker.ietf.org/doc/html/rfc2181#section-7.2) clarified that "This is mentioned nowhere else, and has not generally been implemented." In any case, Amazon's Route 53 uses proactive pushes as a result of `ChangeResourceRecordSets` API calls rather than polling to update records, so it effectively ignores the SOA TTL in zone management.

DNS resolvers do not generally query for the SOA, but they get a copy of it along with a negative hit (`NXDOMAIN` or the newer `NODATA`) to indicate for what zone the negative hit is authoritative. This means, in practice, that it serves as an upper limit on how long a negative hit can be cached. As a result [RFC 2308](https://datatracker.ietf.org/doc/html/rfc2308#section-3) requires that the TTL of a negative answer be set to "_the minimum of_ the MINIMUM field of the SOA record and the TTL of the SOA itself" (emphasis added).

So for consistency, and because it is a separate argument to the `aws_route53_record` resource, we define the negative hit cache TTL via the SOA TTL rather than the SOA `MINIMUM` field, which we have always hard coded (and with this PR, still hard code) as `86400` seconds.


The TTL for NS records is better defined, and has more consequences, but among the key concerns we have is that we do change name servers. In particular, we create separate delegated zones under a master domain name with separate name servers for each environment, and we want those to come up quickly. 

In general, our development environments are so low traffic that any costs associated with short cache times are well worth the benefits in productivity.  That said, more reasonable values for production TTLs would be `300` for SOA and `86400` for NS, but the NS one is particular is hard to be general about, because it depends on your disaster recovery and DDoS mitigation plans. You might want a longer one so that resolvers can serve out of their cache during an outage, or you might want a short one so you can switch DNS servers when one is under attack.

File: variables.tf [hunk 8]
```diff
@@ -21,3 +21,23 @@ variable "parent_zone_record_enabled" {
...
+}
+
+variable "soa_record_ttl" {
+  type        = number
+  default     = 30
+  description = <<-EOT
+  The time to live (TTL) of the start of authority Route53 record, in seconds.
+  
+  The default value for AWS-managed SOA records is `900`, but Cloud Posse prefers shorter TTLs as they result in better availability, despite not being optimal for caching.
+  EOT
```
Comment (Nuru) [inline variables.tf:42]:
```suggestion
  description = <<-EOT
    The time to live (TTL) of the Start of Authority Route53 record, in seconds.
    This sets the maximum time a negative (no data) query can be cached.
    
    The default value is short for responsiveness to changes during development and 
    is not recommended for production. Typical production values are in the range of 300 to 3600.
    EOT
```

File: variables.tf [hunk 9]
```diff
@@ -21,3 +21,23 @@ variable "parent_zone_record_enabled" {
...
 }
+
+variable "ns_record_ttl" {
+  type        = number
+  default     = 60
+  description = <<-EOT
+  The time to live (TTL) of the nameserver Route53 record, in seconds.
+  
+  The default value for AWS-managed NS records is `172800`, but Cloud Posse prefers shorter TTLs as they result in better availability, despite not being optimal for caching.
+  EOT
```
Comment (Nuru) [inline variables.tf:32]:
```suggestion
  description = <<-EOT
    The time to live (TTL) of the Name Server Route53 record, in seconds.
  
    The default value is short for responsiveness to changes during development and 
    is not recommended for production. Typical production values are 86400 or 172800.
    EOT
```

**Review by Nuru — DISMISSED** [review]
I requested changes to the descriptions of the new TTL variables, but otherwise I approve this PR.

**Comment by razorsedge** [comment]
I'm happy so long as I can change the values.

Although I am surprised to learn that CloudPosse module defaults are not tuned for production use.  That should probably be clarified in the module documentation (in all modules).

**Comment by bartelemi** [comment]
can we get this merged?

**Comment by kevcube** [comment]
> can we get this merged?

@bartelemi if you are not in the SweetOps public slack I recommend joining, #pr-reviews in there has proven to be the best place to ask for review and merge attention.

File: examples/complete/variables.tf [hunk 10]
```diff
@@ -5,5 +5,15 @@ variable "region" {
 
 variable "parent_zone_name" {
   type        = string
-  description = "Parent zone name"
+  description = "Name of the hosted zone to contain this record (or specify `parent_zone_id`)"
+}
+
+variable "ns_record_ttl" {
+  type        = number
+  description = "The time to live (TTL) of the nameserver Route53 record, in seconds."
```
Comment (nitrocode) [inline examples/complete/variables.tf:13]: Keep existing default for backwards compatibility

```suggestion
  description = "The time to live (TTL) of the nameserver Route53 record, in seconds."
  default     = "60"
```

File: examples/complete/variables.tf [hunk 11]
```diff
@@ -5,5 +5,15 @@ variable "region" {
...
+}
+
+variable "soa_record_ttl" {
+  type        = number
+  description = "The time to live (TTL) of the start of authority Route53 record, in seconds."
```
Comment (nitrocode) [inline examples/complete/variables.tf:18]: Keep existing default for backwards compatibility

```suggestion
  description = "The time to live (TTL) of the start of authority Route53 record, in seconds."
  default     = "30"
```

File: examples/complete/variables.tf (hunk 10 above)
Comment (nitrocode) [reply examples/complete/variables.tf:13]: Oh this is for the example and you're already providing a value.

File: examples/complete/variables.tf (hunk 11 above)
Comment (nitrocode) [reply examples/complete/variables.tf:18]: Closing. See above.

**Review by nitrocode — DISMISSED** [review]

**Review by nitrocode — APPROVED** [review]

**Pull request merged**